pkg/store/storepb/rules_custom.go

```diff
@@ -0,0 +1,132 @@
+// Copyright (c) The Thanos Authors.
+// Licensed under the Apache License 2.0.
+
+package storepb
+
+import (
+	"encoding/json"
+	"sort"
+	"strconv"
+	"strings"
+
+	"github.com/pkg/errors"
+	"github.com/prometheus/prometheus/pkg/labels"
+)
+
+const (
+	RuleRecordingType = "recording"
+	RuleAlertingType  = "alerting"
+)
+
+func (x *PartialResponseStrategy) UnmarshalJSON(entry []byte) error {
+	fieldStr, err := strconv.Unquote(string(entry))
+	if err != nil {
+		return errors.Wrapf(err, "partialResponseStrategy: unquote %v", string(entry))
+	}
+
+	if len(fieldStr) == 0 {
+		// Default.
+		*x = PartialResponseStrategy_WARN
+		return nil
+	}
+
+	strategy, ok := PartialResponseStrategy_value[strings.ToUpper(fieldStr)]
+	if !ok {
+		return errors.Errorf("unknown partialResponseStrategy: %v", string(entry))
+	}
+	*x = PartialResponseStrategy(strategy)
+	return nil
+}
+
+func (x *PartialResponseStrategy) MarshalJSON() ([]byte, error) {
+	return []byte(strconv.Quote(x.String())), nil
+}
+
+func (m *Rule) UnmarshalJSON(entry []byte) error {
+	decider := struct {
+		Type string `json:"type"`
+	}{}
+	if err := json.Unmarshal(entry, &decider); err != nil {
+		return errors.Wrapf(err, "rule: type field unmarshal: %v", string(entry))
+	}
+
+	switch strings.ToLower(decider.Type) {
+	case "recording":
+		r := &RecordingRule{}
+		if err := json.Unmarshal(entry, r); err != nil {
+			return errors.Wrapf(err, "rule: recording rule unmarshal: %v", string(entry))
+		}
+
+		m.Result = &Rule_Recording{Recording: r}
+	case "alerting":
+		r := &Alert{}
+		if err := json.Unmarshal(entry, r); err != nil {
+			return errors.Wrapf(err, "rule: alerting rule unmarshal: %v", string(entry))
+		}
+
+		m.Result = &Rule_Alert{Alert: r}
+	case "":
+		return errors.Errorf("rule: no type field provided: %v", string(entry))
+	default:
+		return errors.Errorf("rule: unknown type field provided %s; %v", decider.Type, string(entry))
```

Comment:
a bummer that this isn't supported natively, but this looks actually quite ok!

Reply:
I guess it would work for int not string field out of the box

Reply:
Wait, I thought we are talking about enums. Then you probably mean https://github.com/gogo/protobuf/issues/623? I think there will be never native support due to custom `type` thing in JSON (unless it will become a standard).